### PR TITLE
Fix add arithmetic to pass dectest ddAdd.dectest 

### PR DESCRIPTION
**Pull request by Joshcarp**
Passes all halfUp rounding attribute tests in ddAdd.dectest
Does not pass special input tests, weighted Nans and signed Nans

### CODE REVIEW

File: decimal64.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ const (
 
 func signalNaN64() Decimal64 {
 	// TODO: What's the right behavior?
-	panic("sNaN64")
+	return SNaN64
```
Comment (Joshcarp) [inline decimal64.go]: Still haven't found a suitable way to signal the calling function without panicking

File: decimal64.go [hunk 2]
```diff
@@ -102,21 +102,23 @@ func renormalize(exp int, significand uint64) (int, uint64) {
...
+	if exp > mag {
+		return 0, targetExp
 	}
-	for ; exp < -3 && divisor < significand; exp += 4 {
-		divisor *= 10000
-	}
-	for ; exp < -1 && divisor < significand; exp += 2 {
-		divisor *= 100
-	}
-	for ; exp < 0 && divisor < significand; exp++ {
```
Comment (Joshcarp) [inline decimal64.go]: So instead of iterating though exp, which isn't bounded by the significand, we can return zero early if the exp is larger than the magnitude of the exponent, 
Then the powers of 10 global variable can be used with confidence that there isn't going to be any indexing errors

File: decimal64_test.go [hunk 3]
```diff
@@ -50,7 +51,6 @@ func TestDecimal64Float64(t *testing.T) {
 	require.InEpsilon(1.0, one.Float64(), 0.00000001)
 
 	require.True(math.IsNaN(QNaN64.Float64()))
-	require.Panics(func() { SNaN64.Float64() })
```
Comment (Joshcarp) [inline decimal64_test.go]: Snan no longer panics, so tests removed

File: decimal64marshal_test.go [hunk 4]
```diff
@@ -16,7 +16,8 @@ func TestDecimal64Unmarshal(t *testing.T) {
 	require.Equal(t, NewDecimal64FromInt64(23456), d)
 }
 
-func TestDecimal64UnmarshalBadInput(t *testing.T) {
-	var d Decimal64
-	require.Error(t, d.UnmarshalText([]byte("omg")))
-}
+// TODO: find out what proper behaviour is with invalid input
```
Comment (Joshcarp) [inline decimal64marshal_test.go]: Current behaviour is just to return QNan

File: decimal64math.go [hunk 5]
```diff
@@ -27,21 +29,29 @@ func (d Decimal64) Add(e Decimal64) Decimal64 {
...
+	} else {
+		significand = significand2 - significand1
+		exp = exp2
+		sign = sign2
+	}
+	if significand == 0 {
+		return zeroes[sign]
 	}
-	return newFromParts(sign2, exp2, significand2-significand1)
+	exp, significand = renormalize(exp, significand)
```
Comment (Joshcarp) [inline decimal64math.go]: This means that all deicmal64s that represent the same "number" will have the same bit representation

This seemed faster than adding a clause in the Cmp function to match the numbers scales but will mean external number representations will need to be normalised to be used in our library

File: decimal64math.go [hunk 6]
```diff
@@ -7,41 +7,76 @@ func (d Decimal64) Abs() Decimal64 {
 
 // Add computes d + e
 func (d Decimal64) Add(e Decimal64) Decimal64 {
-	flavor1, sign1, exp1, significand1 := d.parts()
-	flavor2, sign2, exp2, significand2 := e.parts()
-	if flavor1 == flSNaN || flavor2 == flSNaN {
+	dDec := d.decParts()
+	eDec := e.decParts()
```
Comment (anzdaddy) [inline decimal64math.go]: nit: `dec` is the adjective of `decParts`. Better to call them `dParts` and `eParts` (or just `ep` and `dp`, given their narrow scope).

File: decimal64math.go [hunk 7]
```diff
@@ -7,41 +7,76 @@ func (d Decimal64) Abs() Decimal64 {
...
+		}
+	} else if dDec.sign == 1 {
+		ans.sign = 1
+		dDec.sign, eDec.sign = 0, 0
+	} else {
+		ans.sign = 0
 	}
-	if significand1 > significand2 {
-		return newFromParts(sign1, exp1, significand1-significand2)
+	if eDec.sign == 0 {
```
Comment (anzdaddy) [inline decimal64math.go]: A little extra math is usually cheaper than a branch.
```go
ans.significand = dDec.significand + (1-2*eDec.sign)*eDec.significand
```

File: decimal64scan_test.go [hunk 8]
```diff
@@ -38,19 +38,19 @@ func TestParseDecimal64Inf(t *testing.T) {
...
 }
 
+// TODO: Fund out what the correct behavior is with bad inputs
+// TODO: Does nan get returned if there are leading/trailing whitespaces?
 func TestParseDecimal64BadInputs(t *testing.T) {
+	require := require.New(t)
 	for _, input := range []string{
 		"", " ", "x",
-		" 0", "0 ",
+		//" 0", "0 ",
```
Comment (anzdaddy) [inline decimal64scan_test.go]: Remove or `// TODO`.

File: decimal64math_test.go [hunk 9]
```diff
@@ -290,9 +276,9 @@ func TestDecimal64SqrtNeg(t *testing.T) {
 
 func TestDecimal64SqrtNaN(t *testing.T) {
 	require := require.New(t)
-
 	require.Equal(QNaN64, QNaN64.Sqrt())
-	require.Panics(func() { SNaN64.Sqrt() })
+	require.Equal(SNaN64, SNaN64.Sqrt())
+
```
Comment (anzdaddy) [inline decimal64math_test.go]: ?

File: decimal64math_test.go [hunk 10]
```diff
@@ -396,6 +382,7 @@ func TestQuoOverflow(t *testing.T) {
 	require.Equal(NegInfinity64, MustParseDecimal64("-1e384").Quo(MustParseDecimal64("0")))
 	require.Equal(QNaN64, Zero64.Quo(Zero64))
 	require.Equal(Zero64, Zero64.Quo(MustParseDecimal64("100")))
+
```
Comment (anzdaddy) [inline decimal64math_test.go]: ?

File: decimal64math_test.go [hunk 11]
```diff
@@ -406,4 +393,5 @@ func TestMul(t *testing.T) {
 	require.Equal(NegZero64, MustParseDecimal64("-1e384").Mul(Zero64))
 	require.Equal(Zero64, Zero64.Mul(Zero64))
 	require.Equal(Zero64, Zero64.Mul(MustParseDecimal64("100")))
+
```
Comment (anzdaddy) [inline decimal64math_test.go]: ?

File: decimalSuite_test.go [hunk 12]
```diff
@@ -20,31 +21,21 @@ type testCaseStrings struct {
...
+		dec64vals := convertToDec64(testVal)
+		testErr := runTest(dec64vals, testVal)
+		if testErr != nil {
+			fmt.Println(testErr)
+			if dec64vals.parseError != nil {
+				fmt.Println(dec64vals.parseError)
+			}
+		}
+	}
+
```
Comment (anzdaddy) [inline decimalSuite_test.go]: Unnecessary blank line

File: decimal64.go [hunk 13]
```diff
@@ -20,9 +41,60 @@ const (
...
+		dec.roundingFunc = roundHalfUp
+	case rDown:
+		dec.roundingFunc = roundDefault
+	case rFloor:
+		dec.roundingFunc = roundDefault
+	case rCeiling:
+		dec.roundingFunc = roundZeroUp
+	case rHalfEven:
+		dec.roundingFunc = roundDefault
+
```
Comment (anzdaddy) [inline decimal64.go]: Unnecessary blank line

File: decimal64.go [hunk 14]
```diff
@@ -177,6 +279,37 @@ func (d Decimal64) parts() (fl flavor, sign int, exp int, significand uint64) {
 	return
 }
 
+func (d Decimal64) decParts() (dP decParts) {
+	dP.sign = int(d.bits >> 63)
```
Comment (anzdaddy) [inline decimal64.go]: Used named return values sparingly.
```go
dP := decParts{sign: …}
```

File: decimal64math.go [hunk 15]
```diff
@@ -7,41 +7,76 @@ func (d Decimal64) Abs() Decimal64 {
...
-
-	if sign1 == sign2 {
-		significand := significand1 + significand2
-		exp1, significand = renormalize(exp1, significand)
-		if significand > maxSig || exp1 > expMax {
-			return infinities[sign1]
+	var ans decParts
+	ans.setRounding(rHalfUp)
+	ans.matchScales(&dDec, &eDec)
+	if eDec.sign != dDec.sign {
```
Comment (anzdaddy) [inline decimal64math.go]: This whole blocks feels like it has too many edge cases.

File: decimal64.go [hunk 16]
```diff
@@ -102,34 +174,63 @@ func renormalize(exp int, significand uint64) (int, uint64) {
...
+	expDiff := targetExp - exp
+	divisor := powersOf10[expDiff]
+	resizedSig := significand / divisor
+	truncatedSig := resizedSig * divisor
+	return significand-truncatedSig >= powersOf10[expDiff-1]
+}
+
+// match scales matches the exponents of d and e and uses dec to get the rounding status
+// decs rounding method must be set or panic will occur
+func (dec *decParts) matchScales(d, e *decParts) {
```
Comment (anzdaddy) [inline decimal64.go]: `dec` is being used merely for its rounding context, and it seems that this is the only use of the relevant fields in `decParts`. I think you can just pass that context in and remove rounding fields from `decParts`.

File: decimal64math_test.go (hunk 10 above)
Comment (Joshcarp) [reply decimal64math_test.go]: should be fixed now

File: decimal64.go [hunk 17]
```diff
@@ -70,6 +75,91 @@ var powersOf10 = []uint64{
...
+		return dec.significand + 1
+	case roundDown:
+		return dec.significand
+	case roundFloor:
+		return dec.significand
+	case roundCeiling: //TODO: fine tune ceiling, floor rounding
+		if rnd.rndStatus == ZR {
+			return dec.significand
+		}
+		if rnd.rndStatus != ZR && dec.sign == 0 {
```
Comment (anzdaddy) [inline decimal64.go]: The second and third `if` statements should be in an else of the first one.

File: decimal64.go [hunk 18]
```diff
@@ -305,3 +473,30 @@ func (d Decimal64) Sign() int {
...
+	if !(flavour1 == flQNaN || flavour1 == flSNaN) {
+		return false
+	}
+	if !(flavour2 == flQNaN || flavour2 == flSNaN) {
+		return true
+	}
+	if flavour1 == flQNaN && flavour2 == flSNaN {
+		return false
+	}
+	return true
```
Comment (anzdaddy) [inline decimal64.go]: A lot of cases. Can't be simpler?

File: decimal64math_test.go (hunk 10 above)
Comment (anzdaddy) [reply decimal64math_test.go]: I'm still seeing a new blank line.

File: decimalSuite_test.go [hunk 19]
```diff
@@ -18,34 +19,63 @@ type testCaseStrings struct {
...
+// 	fmt.Println(g.Add(c))
+// 	require.Equal(a, ans)
 // }
 
+// TODO(joshcarp): This test cannot fail. Proper assertions will be added once the whole suite passes
+// TestFromSuite is the master tester for the dectest suite.
+func TestFromSuite(t *testing.T) {
+	if RUNSUITES {
+		for _, i := range tests {
+			if TESTDEBUG == true {
```
Comment (anzdaddy) [inline decimalSuite_test.go]: Remove ` == true`.

File: decimalSuite_test.go [hunk 20]
```diff
@@ -18,34 +19,63 @@ type testCaseStrings struct {
...
+				fmt.Println("starting test:", i)
+			}
+			testVals := getInput(i)
+			for _, testVal := range testVals {
+				dec64vals := convertToDec64(testVal)
+				testErr := runTest(dec64vals, testVal)
+				// fmt.Println("running test", testVal.testName)
+				if testErr != nil {
+
```
Comment (anzdaddy) [inline decimalSuite_test.go]: Unnecessary blank line

File: decimal64.go [hunk 21]
```diff
@@ -5,46 +5,51 @@ import (
...
+type roundValue int
 
+// roundVal gives information about what the discarded digits were
+// LT -> less than 5, EQ -> Equal to 5, GT -> greater than 5, ZR -> zero, US -> Unset
+const (
+	UNSET roundValue = iota
+	ZR
+	LT
+	EQ
+	GT
```
[the remark this answers is not in the document]
Comment (Joshcarp) [reply decimal64.go]: yep

File: decimal64.go (hunk 18 above)
Comment (Joshcarp) [reply decimal64.go]: is simpler now

File: decimalSuite_test.go (hunk 19 above)
Comment (Joshcarp) [reply decimalSuite_test.go]: I need to write this down somewhere

File: decimal64.go (hunk 17 above)
Comment (Joshcarp) [reply decimal64.go]: Fixed but also commented out roundCeiling; not working correctly yet

File: decimal64.go [hunk 22]
```diff
@@ -70,6 +77,88 @@ var powersOf10 = []uint64{
...
+		// dec.exp--
+		// return dec.significand - 1
+	case roundHalfEven:
+		if rnd.rndStatus == lt || rnd.rndStatus == zr {
+			return dec.significand
+		}
+		if rnd.rndStatus == eq && dec.significand%2 == 0 {
+			return dec.significand
+		}
+		return dec.significand + 1
```
Comment (anzdaddy) [inline decimal64.go]: Replace the entire `roundHalfEven` case with:
```go
if rnd.rndStatus != lt && rnd.rndStatus != zr && (rnd.rndStatus != eq || dec.significand%2 != 0) {
	return dec.significand + 1
}
```

File: decimal64.go [hunk 23]
```diff
@@ -70,6 +77,88 @@ var powersOf10 = []uint64{
...
+	switch rnd.roundType {
+	case roundHalfUp:
+		if rnd.rndStatus == lt || rnd.rndStatus == zr {
+			return dec.significand
+		}
+		return dec.significand + 1
+	case roundDown:
+		return dec.significand
+	case roundFloor:
+		return dec.significand
```
Comment (anzdaddy) [inline decimal64.go]: `roundDown` and `roundFloor` cases match the default behaviour. Remove them.

File: decimal64.go [hunk 24]
```diff
@@ -70,6 +77,88 @@ var powersOf10 = []uint64{
...
+func (dec *decParts) round(rnd roundContext) uint64 {
+	if rnd.rndStatus == unset {
+		return dec.significand
+	}
+	switch rnd.roundType {
+	case roundHalfUp:
+		if rnd.rndStatus == lt || rnd.rndStatus == zr {
+			return dec.significand
+		}
+		return dec.significand + 1
```
Comment (anzdaddy) [inline decimal64.go]: Invert the test and leave out `return dec.significand`.

File: decimal64.go [hunk 25]
```diff
@@ -102,34 +191,66 @@ func renormalize(exp int, significand uint64) (int, uint64) {
...
+	dec.significand = dec.significand / divisor
+	dec.exp = targetExp
+	return
+}
+
+// roundStatus gives info abour the truncated part of the significand that can't be fully stored in 16 decimal digits.
+func roundStatus(significand uint64, exp int, targetExp int) roundValue {
+	expDiff := targetExp - exp
+	if expDiff > 19 {
+		if significand != 0 {
```
Comment (anzdaddy) [inline decimal64.go]: Fold the two `if`s into one.

File: decimal64.go [hunk 26]
```diff
@@ -102,34 +191,66 @@ func renormalize(exp int, significand uint64) (int, uint64) {
...
+			return lt
+		}
 	}
-	for ; exp < -3 && divisor < significand; exp += 4 {
-		divisor *= 10000
+	divisor := powersOf10[expDiff]
+	resizedSig := significand / divisor
+	truncatedSig := resizedSig * divisor
+	if significand-truncatedSig == 0 {
```
Comment (anzdaddy) [inline decimal64.go]: Too much branching. Maybe a switch works better here?
```go
switch (significand-truncatedSig)/powersOf10[expDiff-1] {
case 0:
	return zr
case 1, 2, 3, 4:
	return lt
case 5:
	return eq
default:
	return gt
}
```

File: decimalSuite_test.go [hunk 27]
```diff
@@ -18,34 +19,62 @@ type testCaseStrings struct {
...
+// 	// g := MustParseDecimal64("sNaN")
+// 	fmt.Println(g.Add(c))
+// 	require.Equal(a, ans)
 // }
 
+// TODO(joshcarp): This test cannot fail. Proper assertions will be added once the whole suite passes
+// TestFromSuite is the master tester for the dectest suite.
+func TestFromSuite(t *testing.T) {
+	if RUNSUITES {
+		for _, i := range tests {
```
Comment (anzdaddy) [inline decimalSuite_test.go]: `i` is normally thought of as an integer. Better to call it `test` (or `file`, after `getInput()`'s parameter).

File: decimal64.go [hunk 28]
```diff
@@ -305,3 +472,28 @@ func (d Decimal64) Sign() int {
...
+	numBits := 64 - bits.LeadingZeros64(n)
+	numDigits := numBits * 3 / 10
+	if n < powersOf10[numDigits] {
+		return numDigits
+	}
+	return numDigits + 1
+}
+
+// propogateNan returns true if flavor1 should be propogated else false if flavor2 should be propogated
+func propogateNan(flavour1, flavour2 flavor) bool {
```
[the remark this answers is not in the document]
Comment (Joshcarp) [reply decimal64.go]: fixed it by adding a pointer field to decParts that points to the original decimal part, then the propogateNan function can return the correct decimal without needing to return a bool or flavor but the address to the original nan

File: decimal64.go [hunk 29]
```diff
@@ -5,13 +5,19 @@ import (
...
+type roundValue int
+
+// roundVal gives information about what the discarded digits were
+// LT -> less than 5, EQ -> Equal to 5, GT -> greater than 5, zr -> zero, US -> unset
+const (
+	unset roundValue = iota
+	zr
+	lt
+	eq
+	gt
```
[the remark this answers is not in the document]
Comment (Joshcarp) [reply decimal64.go]: going to omit dd as it'd be the same for other deicmal types

File: decimal64.go [hunk 30]
```diff
@@ -486,17 +488,15 @@ func size(n uint64) int {
 
 // propogateNan returns true if flavor1 should be propogated else false if flavor2 should be propogated
 func propogateNan(flavour1, flavour2 flavor) bool {
-	if flavour1 == flSNaN || flavour2 == flSNaN {
+	if flavour1 == flSNaN {
 		signalNaN64()
+		return true
```
[the remark this answers is not in the document]
Comment (Joshcarp) [reply decimal64.go]: Signal Nan does not return SNan64 anymore, needed for weighted Nans (next PR) hopefully...

File: decimal64.go [hunk 31]
```diff
@@ -483,17 +463,17 @@ func size(n uint64) int {
 	return numDigits + 1
 }
 
-// propogateNan returns true if flavor1 should be propogated else false if flavor2 should be propogated
-func propogateNan(flavour1, flavour2 flavor) bool {
-	if flavour1 == flSNaN {
+// propagateNan returns true if flavor1 should be propagated else false if flavor2 should be propagated
+func propagateNan(dp, ep *decParts) *Decimal64 {
+	if dp.fl == flSNaN {
```
Comment (anzdaddy) [inline decimal64.go]:
```go
switch dp.fl {
```

File: decimal64.go [hunk 32]
```diff
@@ -5,13 +5,17 @@ import (
...
-type Decimal64 struct {
-	bits uint64
-}
-
 type flavor int
+type roundingMode int
+type discardedDigit int
+
+const (
+	unset discardedDigit = iota
```
Comment (anzdaddy) [inline decimal64.go]: `unset` is too generic. Need to related the name to `discardedDigits` somehow. I'd be happy with something verbose like `unsetDiscardedDigit`, since it's never actually used. A perfectly cromulent alternative is to start with `eq0 discardedDigit = 1 + iota`.

File: decimal64.go [hunk 33]
```diff
@@ -176,6 +275,45 @@ func (d Decimal64) parts() (fl flavor, sign int, exp int, significand uint64) {
 	}
 	return
 }
+func (dec decParts) isNan() bool {
+	return dec.fl == flQNaN || dec.fl == flSNaN
+}
+
+// decParts gets the parts and returns in decParts stuct, doesn't get the magnitude due to performance issues
+func (d *Decimal64) getParts() decParts {
```
Comment (anzdaddy) [inline decimal64.go]: The previous name was fine. Just `parts` would be even better.

soapbox: I'm not a fan of `get…` prefixes, since such functions generally don't mutate their receiver. They don't do something to the subject; they simply tell you something about it. From another perspective, Calling it `getParts` instead of `parts` is like calling the trigonometry function `computeCos` instead of just `cos`.

File: decimal64.go [hunk 34]
```diff
@@ -305,3 +447,28 @@ func (d Decimal64) Sign() int {
...
+	numDigits := numBits * 3 / 10
+	if n < powersOf10[numDigits] {
+		return numDigits
+	}
+	return numDigits + 1
+}
+
+// propagateNan returns true if flavor1 should be propagated else false if flavor2 should be propagated
+func propagateNan(dp, ep *decParts) *Decimal64 {
+	if dp.fl == flSNaN {
```
Comment (anzdaddy) [inline decimal64.go]:
```go
switch dp.fl {
```

File: decimal64.go [hunk 35]
```diff
@@ -305,3 +447,28 @@ func (d Decimal64) Sign() int {
...
+		signalNaN64()
+		return dp.dec
```
Comment (anzdaddy) [inline decimal64.go]: This line never executes.

File: decimal64math.go [hunk 36]
```diff
@@ -7,41 +7,66 @@ func (d Decimal64) Abs() Decimal64 {
...
 			return e
 		}
-		if flavor2 != flInf || sign1 == sign2 {
+		if ep.fl != flInf || ep.sign == dp.sign {
 			return d
 		}
 		return QNaN64
 	}
-	if significand1 == 0 && significand2 == 0 {
+	if ep.significand == 0 || dp.significand == 0 {
```
Comment (anzdaddy) [inline decimal64math.go]: Suggestion (though I haven't considered rules for ±0):
```go
if ep.significand == 0 {
	return d
}
if dp.significant == 0 {
	return e
}
```

File: decimal64math.go [hunk 37]
```diff
@@ -52,20 +77,17 @@ func (d Decimal64) Add(e Decimal64) Decimal64 {
...
 //
 func (d Decimal64) Cmp(e Decimal64) int {
-	flavor1, _, _, _ := d.parts()
-	flavor2, _, _, _ := e.parts()
+	flavor1, _, _, significand1 := d.parts()
+	flavor2, _, _, significand2 := e.parts()
 	if flavor1 == flSNaN || flavor2 == flSNaN {
 		signalNaN64()
-		return 0
+		return -2
```
Comment (anzdaddy) [inline decimal64math.go]: This line is never reached.

File: decimal64math.go [hunk 38]
```diff
@@ -78,22 +100,20 @@ func (d Decimal64) Cmp(e Decimal64) int {
 func (d Decimal64) Mul(e Decimal64) Decimal64 {
 	flavor1, sign1, exp1, significand1 := d.parts()
 	flavor2, sign2, exp2, significand2 := e.parts()
-
 	if flavor1 == flSNaN || flavor2 == flSNaN {
-		return signalNaN64()
+		signalNaN64()
+		return SNaN64
```
Comment (anzdaddy) [inline decimal64math.go]: This line is never reached.

File: decimal64math.go [hunk 39]
```diff
@@ -112,14 +132,13 @@ func (d Decimal64) Neg() Decimal64 {
 func (d Decimal64) Quo(e Decimal64) Decimal64 {
 	flavor1, sign1, exp1, significand1 := d.parts()
 	flavor2, sign2, exp2, significand2 := e.parts()
-
 	if flavor1 == flSNaN || flavor2 == flSNaN {
-		return signalNaN64()
+		signalNaN64()
+		return SNaN64
```
Comment (anzdaddy) [inline decimal64math.go]: This line is never reached.

File: decimal64math.go [hunk 40]
```diff
@@ -163,10 +180,10 @@ func (d Decimal64) Sqrt() Decimal64 {
 	case flQNaN:
 		return d
 	case flSNaN:
-		return signalNaN64()
+		signalNaN64()
+		return SNaN64
```
Comment (anzdaddy) [inline decimal64math.go]: This line is never reached.

File: decimal64scan.go [hunk 41]
```diff
@@ -36,6 +36,7 @@ func MustParseDecimal64(s string) Decimal64 {
 
 // Scan implements fmt.Scanner.
 func (d *Decimal64) Scan(state fmt.ScanState, verb rune) error {
+	*d = SNaN64
```
Comment (anzdaddy) [inline decimal64scan.go]: Don't change it if returning an `error`.

File: decimal64scan_test.go [hunk 42]
```diff
@@ -38,19 +39,18 @@ func TestParseDecimal64Inf(t *testing.T) {
 	parseEquals64(QNaN64, "NaN")
 }
 
+// TODO: Fund out what the correct behavior is with bad inputs
```
Comment (anzdaddy) [inline decimal64scan_test.go]: Typo

File: decimalSuite_test.go [hunk 43]
```diff
@@ -120,14 +147,19 @@ func runTest(testVals decValContainer, testValStrings testCaseStrings) error {
...
 }
 
 // TODO: get runTest to run more functions such as FMA.
 // execOp returns the calculated answer to the operation as Decimal64.
 func execOp(val1, val2 Decimal64, op string) Decimal64 {
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("Recovered from", r)
+		}
+	}()
```
Comment (anzdaddy) [inline decimalSuite_test.go]: Why is this needed?

File: decimal64.go (hunk 31 above)
Comment (Joshcarp) [reply decimal64.go]:
``` switch dp.fl{

... 

switch ep.fl{

```

don't see how this is going to make it any more clear

File: decimal64.go (hunk 32 above)
Comment (Joshcarp) [reply decimal64.go]: I'm just going to get rid of unset
as you said, it doesn't actually get used apart from initialisation.

File: decimal64.go (hunk 33 above)
Comment (Joshcarp) [reply decimal64.go]: Yes
Although previous name was ambiguous with struct
I will eventually name this function parts but there's still the original parts function that is still a dependency for other functions

File: decimalSuite_test.go (hunk 43 above)
Comment (Joshcarp) [reply decimalSuite_test.go]: to stop the testing function from crashing when a Nan is signalled (from signalNaN64())

File: decimal64scan.go (hunk 41 above)
Comment (Joshcarp) [reply decimal64scan.go]: default behaviour from any bad input is SNaN64 though which is why I set the pointer at the start

File: decimal64.go (hunk 35 above)
Comment (Joshcarp) [reply decimal64.go]: I've defered panic now

File: decimal64math.go (hunk 37 above)
Comment (Joshcarp) [reply decimal64math.go]: deferred

File: decimal64math.go (hunk 38 above)
Comment (Joshcarp) [reply decimal64math.go]: deferred

File: decimal64math.go (hunk 39 above)
Comment (Joshcarp) [reply decimal64math.go]: deferred

File: decimal64.go (hunk 31 above)
Comment (Joshcarp) [reply decimal64.go]: changed anyway

File: decimal64.go (hunk 33 above)
Comment (Joshcarp) [reply decimal64.go]: added TODO

File: decimal64.go (hunk 34 above)
Comment (Joshcarp) [reply decimal64.go]: behaviour between sNans and qNans are the same now

File: decimal64math.go (hunk 39 above)
Comment (Joshcarp) [reply decimal64math.go]: panics removed

File: decimalSuite_test.go (hunk 43 above)
Comment (Joshcarp) [reply decimalSuite_test.go]: removed

File: decimal64math.go (hunk 38 above)
Comment (Joshcarp) [reply decimal64math.go]: panics removed

File: decimal64.go [hunk 44]
```diff
@@ -305,3 +444,21 @@ func (d Decimal64) Sign() int {
 func (d Decimal64) Signbit() bool {
 	return d.bits>>63 == 1
 }
+
+//size returns the magnitude (number of digits) of a uint64.
+func size(n uint64) int {
```
Comment (anzdaddy) [inline decimal64.go]: nit: Needs a better name. Suggestion: `numDecimalDigits` (shorter would be better, but I can't think of anything off the top of my head).

File: decimal64math.go [hunk 45]
```diff
@@ -7,41 +7,63 @@ func (d Decimal64) Abs() Decimal64 {
...
-
-	if sign1 == sign2 {
-		significand := significand1 + significand2
-		exp1, significand = renormalize(exp1, significand)
-		if significand > maxSig || exp1 > expMax {
-			return infinities[sign1]
+	ep.updateMag()
+	dp.updateMag()
+	sep := dp.separation(ep)
+	var ans decParts
```
Comment (anzdaddy) [inline decimal64math.go]: Move `ans` down to just before the `if` that fills it.
